File: HackerRank/Booking Hackathon/nearby.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

#define PI 3.14159265359
struct point {
    double latitude, longitude;
};

double degree2radians(double angle) {
    return angle * PI / 180;
}

double distance_between(point point1, point point2) {
    const int EARTH_RADIUS = 6371;
    double point1_lat_in_radians  = degree2radians( point1.latitude );
    double point2_lat_in_radians  = degree2radians( point2.latitude );
    double point1_long_in_radians  = degree2radians( point1.longitude );
    double point2_long_in_radians  = degree2radians( point2.longitude );

    double ret = acos( sin( point1_lat_in_radians ) * sin( point2_lat_in_radians ) +
                 cos( point1_lat_in_radians ) * cos( point2_lat_in_radians ) *
                 cos( point2_long_in_radians - point1_long_in_radians) ) * EARTH_RADIUS;
    return round(ret * 100) / 100;
}

struct attraction {
    int id;
    point loc;
};

int main() {
    int n;
    cin >> n;
    vector<attraction> attractions(n);
    for (int i = 0; i < n; i++) {
        int id;
        double latitude, longitude;
        cin >> id >> latitude >> longitude;
        attractions[i].id = id;
        attractions[i].loc.latitude = latitude;
        attractions[i].loc.longitude = longitude;
    }
    int m;
    cin >> m;
    for (int i = 0; i < m; i++) {
        double latitude, longitude, maxTime;
        string transport;
        cin >> latitude >> longitude >> transport >> maxTime;
        int speed;
        if (transport == "metro") {
            speed = 20;
        } else if (transport == "bike") {
            speed = 15;
        } else {
            speed = 5;
        }
        double maxDist = speed * maxTime / 60.0;
        point loc;
        loc.latitude = latitude;
        loc.longitude = longitude;
        vector<attraction> canReach;
        for (int j = 0; j < n; j++) {
            double dist = distance_between(loc, attractions[j].loc);
            if (dist <= maxDist) {
                canReach.push_back(attractions[j]);
            }
        }
        sort(canReach.begin(), canReach.end(), [&loc](attraction a, attraction b) {
            double distA = distance_between(loc, a.loc);
            double distB = distance_between(loc, b.loc);
            if (distA == distB) {
                return a.id < b.id;
            }
            return distA < distB;
        });
        cout << canReach[0].id;
        for (int i = 1; i < canReach.size(); i++) {
            cout << ' ' << canReach[i].id;
        }
        cout << '\n';
    }
}

File: HackerRank/Booking Hackathon/nearby.hs
import Data.Tuple

pi :: (Floating a) => a
pi = 3.14159265359

data Point a b = Point { lat :: a
                       , long :: b
                       } deriving (Eq, Show, Read)

data Attraction i p = Attraction { _id :: i
                                 , loc :: p
                                 } deriving (Eq, Show, Read)

data Guest p v t = Guest{ loc2 :: p
                        , speed :: v
                        , time :: t
                        } deriving (Eq, Show, Read)

degree2radians :: (Floating f) => f -> f
degree2radians = (* (Main.pi / 180));

distanceBetween :: (Floating f) => Point f f -> Point f f -> f
distanceBetween (Point {lat = lat1, long = long1}) (Point {lat = lat2, long = long2}) =
    let earth_radius = 6371
    in earth_radius * acos((sin . degree2radians) lat1 * (sin . degree2radians) lat2
        + (cos . degree2radians) lat1 * (cos . degree2radians) lat2 * cos(degree2radians long2
        - degree2radians long1))

convertToAttraction :: String -> Attraction Int (Point Double Double)
convertToAttraction line =
    let ws = words line
        __id = read (head ws) :: Int
        _lat = read (ws !! 1) :: Double
        _long = read (last ws) :: Double
        _loc = Point {lat = _lat, long = _long}
    in Attraction {_id = __id, loc = _loc}

convertToAttractions :: [String] -> [Attraction Int (Point Double Double)]
convertToAttractions = map convertToAttraction

convertToGuest :: String -> Guest (Point Double Double) Int Int
convertToGuest line =
    let ws = words line
        _lat = read (head ws) :: Double
        _long = read (ws !! 1) :: Double
        _loc2 = Point {lat = _lat, long = _long}
        transport = ws !! 2
        _speed = if transport == "metro" then 20 else if transport == "bike" then 15 else 5
        _time = read (last ws) :: Int
    in Guest {loc2 = _loc2, speed = _speed, time = _time}

convertToGuests :: [String] -> [Guest (Point Double Double) Int Int]
convertToGuests = map convertToGuest

cmpAttractions :: Attraction Int (Point Double Double) -> Attraction Int (Point Double Double)
        -> Point Double Double -> Bool
cmpAttractions a b start
    | distanceBetween (loc a) start == distanceBetween (loc b) start = (_id a) < (_id b)
    | otherwise = distanceBetween (loc a) start < distanceBetween (loc b) start

sort :: Point Double Double -> [Attraction Int (Point Double Double)] -> [Attraction Int (Point Double Double)]
sort p [] = []
sort p (a:as) =
    let less = sort p [x | x <- as, cmpAttractions x a p]
        more = sort p [x | x <- as, cmpAttractions a x p]
    in  less ++ [a] ++ more

solveOne :: Guest (Point Double Double) Int Int -> [Attraction Int (Point Double Double)] -> [Int]
solveOne g as = let maxDist = (fromIntegral $ speed g) * (fromIntegral $ time g) / 60.0
    in map _id $ sort (loc2 g) $ filter (\a -> distanceBetween (loc2 g) (loc a) <= maxDist) as

solveAll :: [Guest (Point Double Double) Int Int] -> [Attraction Int (Point Double Double)] -> [[Int]]
solveAll (g:gs) as = solveOne g as : solveAll gs as
solveAll [] as = []

getLines :: (Integral a) => a -> IO [String]
getLines n = do
    if n > 0
        then do
            line <- getLine
            _lines <- getLines $ n - 1
            return (line:_lines)
        else
            return []

putStrLines :: [String] -> IO ()
putStrLines (s:ss) = do
    putStrLn s
    if ss == []
        then do
            return ()
        else
            putStrLines ss

main :: IO ()
main = do
    nStr <- getLine
    let n = read nStr :: Int
    attractionData <- getLines n
    let attractions = convertToAttractions attractionData
    mStr <- getLine
    let m = read mStr :: Int
    guestData <- getLines m
    let guests = convertToGuests guestData
    putStrLines $ map (foldr1 (\x y -> x ++ " " ++ y) . map show) $ solveAll guests attractions